### PR TITLE
Pylock

### PR DESCRIPTION
**Pull request by ghost**
This change implements a processing guard that only allows processing of external calls (e.g. python) during CApplication::FrameMove. This change is necessary because of syncronization issues between Renderer (mainthread) and modifications in GUI wich are done from python thread:

Pls. refer to:
https://github.com/xbmc/xbmc/pull/9698

for an example of the issue.

### CODE REVIEW

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -2657,7 +2678,21 @@ void CApplication::FrameMove(bool processEvents, bool processGUI)
...
+
+    // Open the door for external calls e.g python exactly here.
+    // We try to close the door latest after 10 ms.
+    unsigned int maxLoops = 5;
+    m_ProcessedExternalCalls = 0;
+    while ((m_WaitingExternalCalls || m_ProcessedExternalCalls) && maxLoops--)
+    {
+      m_ProcessedExternalCalls = 0;
+      m_frameMoveGuard.unlock();
+      Sleep(2);
```
Comment (FernetMenta) [inline xbmc/Application.cpp]: this window should start at zero. assume there is only a single python call waiting, that only takes some micro secs to be completed. we can expand the window if there are threads waiting for x times in a row when calling FrameMove.

This code blocks the main thread at least for 4ms  :(

File: xbmc/Application.cpp [hunk 2]
```diff
@@ -2622,6 +2628,21 @@ void CApplication::HandleShutdownMessage()
   }
 }
 
+void CApplication::LockFrameMoveGuard()
+{
+  AtomicIncrement(&m_WaitingExternalCalls);
```
Comment (FernetMenta) [inline xbmc/Application.cpp]: please drop this ancient method. c++ supports atomics

File: xbmc/Application.h [hunk 3]
```diff
@@ -509,7 +519,11 @@ class CApplication : public CXBApplicationEx, public IPlayerCallback, public IMs
   bool m_fallbackLanguageLoaded;
   
 private:
-  CCriticalSection                m_critSection;                 /*!< critical section for all changes to this class, except for changes to triggers */
+  CCriticalSection m_critSection;                 /*!< critical section for all changes to this class, except for changes to triggers */
+  
+  CCriticalSection m_frameMoveGuard;              /*!< critical section for synchronizing GUI actions from inside and outside (python) */
+  long m_WaitingExternalCalls;                    /*!< counts threads wich are waiting to be processed in FrameMove */
```
Comment (FernetMenta) [inline xbmc/Application.h]: std::atomic_int

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
jenkins was almost fine: http://jenkins.kodi.tv/view/Linux/job/BuildMulti-PR/4985/
I had to kill Windows build due to usual error: cannot stat ...

**Comment by MartijnKaijser** [comment]
Looks like this PR makes Kodi hang on startup and doesn't go past the splash screen. At for the droid builds i've based on this PR

**Comment by mapfau** [comment]
@MartijnKaijser could you pls provide a log? And best possible something like a backtrace?
Important are the two threads "languageInvoker" and the main thread.

**Comment by MartijnKaijser** [comment]
Tried on two devices and both lock up on splash screen
http://xbmclogs.com/pc0jiiwcx

**Comment by MartijnKaijser** [comment]
Doesn't seem to happen on windows from testing however it for sure does on android as I've made a build before and after.

**Comment by mapfau** [comment]
Thanks, just looked into log but there is no chance to see anything.
I know that there could be places where mainthread and languageInvoker could dead lock, to really see the place I need some backtrace of the two threads.

Are you able to attach wich an debugger, then toggle to both threads and make an back-trace of them? With this information I could easily see whats the reason....

**Comment by FernetMenta** [comment]
looks like Android only issue. in this case we certainly do not revert

**Comment by MartijnKaijser** [comment]
What? If it was linux it would be reverted instantly.

**Comment by FernetMenta** [comment]
we don't revert for a decaying platform without maintainer. this pr is a very important. i suggest to fix that Android crap instead doing harm to the maintained platforms

**Comment by ronie** [comment]
this one is pretty spammy:
https://github.com/xbmc/xbmc/pull/9748/commits/82218b570c8d5e7d2ace70cbe71f24e101e72163#diff-63246197ac0f0ac89358c1155ad21d5bR2689

it's hitting my log ~40 times every minute

**Comment by mapfau** [comment]
If there are no known issues with this "FrameMoveGuard", we can for sure remove the line.
@FernetMenta : anything popped up regarding this?

**Comment by FernetMenta** [comment]
Yes, I think we should drop the this log

**Comment by mapfau** [comment]
I'll push an PR tomorrow

**Pull request merged**